## Number Bases and Math Utilities

This tutorial explores key built-in Python functions for manipulating numbers, including converting between different bases (like binary and hexadecimal) and performing essential mathematical operations.

### The Code Example

Here is the complete snippet we will be breaking down:

In [13]:
# --- 1. Base Conversion Functions (Decimal to Other Bases) ---
print("--- Base Conversion Output ---")
print(f"Decimal 12 in Hex: {hex(12)}")
print(f"Decimal 12 in Octal: {oct(12)}")
print(f"Decimal 12 in Binary: {bin(12)}")
print("-" * 30)

# --- 2. Using Hexadecimal in Range ---
print("--- Range Loop Output (0x16 = 22) ---")
for i in range(0x16):
    print(i, end=' ')
print("\n" + "-" * 30)

# --- 3. Base Conversion Function (String to Decimal) ---
print("--- String-to-Decimal Conversion ---")
# Converts the string '45' from base 8 (octal) into a decimal integer.
print(f"Octal '45' to Decimal: {int('45', 8)}")
print("-" * 30)

# --- 4. Mathematical Utility Functions ---
print("--- Math Utility Output ---")
# Calculates (2 ** 10) % 100 (Power and Modulo)
print(f"pow(2, 10, 100): {pow(2, 10, 100)}")

# Calculates the Absolute value
print(f"abs(-4): {abs(-4)}")

# Rounds the float to the nearest integer
print(f"round(3.14): {round(3.14)}")

--- Base Conversion Output ---
Decimal 12 in Hex: 0xc
Decimal 12 in Octal: 0o14
Decimal 12 in Binary: 0b1100
------------------------------
--- Range Loop Output (0x16 = 22) ---
0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 
------------------------------
--- String-to-Decimal Conversion ---
Octal '45' to Decimal: 37
------------------------------
--- Math Utility Output ---
pow(2, 10, 100): 24
abs(-4): 4
round(3.14): 3


## 🔬  Advanced Numbers - The Decimal Module

Standard Python `float` numbers (like `3.14`) are stored using binary floating-point representation, which can lead to small, unexpected errors in calculations due to precision loss. This tutorial introduces the **`decimal` module** to perform arithmetic with arbitrary precision, which is essential for financial, scientific, and other high-accuracy applications.

### The Problem with Standard Floats

First, let's look at the problem the `decimal` module solves.


In [10]:
# Standard float calculation demonstrating precision error
standard_sum = 0.1 + 0.1 + 0.1
standard_product = 0.1 * 3

# Notice the small error in the output
print("--- Standard Float Issue ---")
print(f"0.1 + 0.1 + 0.1 = {standard_sum}")
print(f"0.1 * 3         = {standard_product}")
print("-" * 30)

--- Standard Float Issue ---
0.1 + 0.1 + 0.1 = 0.30000000000000004
0.1 * 3         = 0.30000000000000004
------------------------------


### The Solution: The `decimal` Module

The `decimal` module stores numbers as actual decimal digits, ensuring exact precision for financial and comparative calculations.

In [11]:
from decimal import Decimal, getcontext, localcontext

# 1. Set the Global Precision (how many digits to track)
# Default is 28. We'll set it higher for demonstration.
getcontext().prec = 50 

# 2. Correct way to create Decimal objects: using strings
# NEVER create Decimal from a float (e.g., Decimal(0.1))
d_sum = Decimal('0.1') + Decimal('0.1') + Decimal('0.1')
d_product = Decimal('0.1') * Decimal(3)

# 3. Performing a complex division with high precision
result_div = Decimal(1) / Decimal(7)

print("--- Decimal Solution ---")
print(f"0.1 + 0.1 + 0.1 = {d_sum}")
print(f"0.1 * 3         = {d_product}")
print(f"1 / 7 (Precision 50) = {result_div}")
print("-" * 30)

# 4. Using the Decimal context manager to change precision locally
with localcontext() as ctx:
    ctx.prec = 4 # Temporarily set precision to 4 digits
    local_div = Decimal(1) / Decimal(3)

print(f"1 / 3 (Local Precision 4) = {local_div}")

--- Decimal Solution ---
0.1 + 0.1 + 0.1 = 0.3
0.1 * 3         = 0.3
1 / 7 (Precision 50) = 0.14285714285714285714285714285714285714285714285714
------------------------------
1 / 3 (Local Precision 4) = 0.3333


### Explanation and Key Concepts

#### 1. Why Standard Floats Fail

When you run `0.1 + 0.1 + 0.1`, the result is `0.30000000000000004`. This is because standard Python floats use the **IEEE 754 standard** to represent numbers in binary (base 2). Just as the fraction $1/3$ cannot be represented perfectly in decimal (it becomes $0.333...$), the fraction $1/10$ cannot be represented perfectly in binary.

#### 2. Importing and Creating Decimals

* **Import:** You must import the `Decimal` class: `from decimal import Decimal`.
* **Initialization (CRITICAL):** You must initialize `Decimal` objects using strings (`Decimal('0.1')`). If you use `Decimal(0.1)`, Python first converts `0.1` to its imperfect binary float representation before the `Decimal` class sees it, defeating the purpose.

#### 3. Setting Precision (Context)

The `decimal` module uses a **Context** to define the rules for arithmetic, including the number of significant digits it will track.

* `getcontext().prec = 50`: This sets the global precision for all subsequent `Decimal` calculations to 50 significant digits (the default is 28). This is why `1 / 7` gives a very long, accurate result.
* `localcontext()`: You can use a context manager to temporarily change the precision for a small block of code without affecting the rest of your program.



In [9]:
print (hex(12))
print (oct(12))
print (bin(12))
for i in range(0x16):
    print(i)

int('45',8)
pow(2,10,100)
abs(-4)
round(3.14)

0xc
0o14
0b1100
0
1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20
21


3